design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,421 @@
+  [BIOS API](https://docs.openstack.org/api-ref/baremetal/?expanded=#node-bios-nodes)
+  will be stored in the Status section along with the BIOS Registry
+  information which provides a schema for the settings.
+- `vendorSettings` - BIOS Settings to write to the host via Ironic will be
+  stored in the Spec section. It will be initially populated with the
+  name/value pairs from `settingsData` that do not have the `read-only` or
+  `unique` schema flag set. Existing values will not be overwritten.
+
+A second CRD `FirmwareSchema` will added to store schema data (aka BIOS
+Registry) received from Ironic via the BIOS API. This schema provides
+additional info about each BIOS setting along with limits that can be used
+both by a user when changing the settings or as a part of input validation.
+The BIOS Registry support was recently added to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681).
+This schema is the same for each vendor's model so it will not need to be
+updated when stored. More info for the BIOS registry data can be found in the
+[Redfish Registry](https://redfish.dmtf.org/schemas/v1/AttributeRegistry.v1_3_5.json)
+
+The schema fields returned via the BIOS API include:
+
+- Type (e.g. enum, string, integer, boolean, or password)
+- ReadOnly (a boolean indicating whether the attribute is read only)
+- Unique (indicating the value is specific to the host where it was
+  retrieved)
+- Allowable values for enum type attributes
+- Minimum and maximum values (for numeric attributes)
+- Minimum and maximum character lengths (for string attributes)
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. The baremetal-operator will manage the data as follows:
+
+- The node first transitions to manageable during the `Registration` state, so
+  at the end of that state `settingsData` will be populated.
+- Settings can be updated during the `Preparing` state, so at the end of that
+  state the settings will also be retrieved and used to update
+  `settingsData`.
+
+Upon retrieving the BIOS Settings from Ironic, the baremetal-operator will
+copy the name/value pairs to the `vendorSettings` field in
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
+
+A user can update `vendorSettings` to the desired values for the
+settings using the limits from the `FirmwareSchema` section as a guide.
+
+The baremetal-operator will detect when `vendorSettings` changes by comparing
+the name/value pairs to `settingsData`. When a change is detected, the BMO
```

Comment:
Do we need to restrict in which BMO states this will be possible? If so, it would be useful to list them here.

Reply:
Clean steps will be updated at the same time its currently done, when building the ManualCleanSteps https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L935 in actionPreparing:
https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L764.

I'll update this section to make it clear.